src/app/GitUI/Editor/Diff/DiffLineNumAnalyzer.cs

```diff
@@ -150,25 +152,60 @@ public static DiffLinesInfo Analyze(TextEditorControl textEditor, bool isCombine
         }
 
         return ret;
-    }
 
-    private static bool IsMinusLine(string line)
-    {
-        return line.StartsWith('-');
-    }
+        static bool IsGitWordMatch(DiffLineType lineType, string line, int textOffset, int textLength, List<TextMarker> textMarkers)
+        {
+            // Heuristics (or wild guessing): For GitWordDiff find if the line is exclusive (otherwise DiffLineType.MinusPlus).
+            // If the marker covers the line this should be true.
+            // Git output is impossible to parse, some guesses are done.
+            // Whitespace only lines are still incorrect (no marker at all in Git),
+            // as well as some other situations.
 
-    private static bool IsPlusLine(string line)
-    {
-        return line.StartsWith('+');
-    }
+            int firstNonWhiteSpace = line.Length - line.AsSpan().TrimStart().Length;
 
-    private static bool IsPlusLineInCombinedDiff(string line)
-    {
-        return line.StartsWith("++") || line.StartsWith("+ ") || line.StartsWith(" +");
-    }
+            return textMarkers.Count == 1
 
-    private static bool IsMinusLineInCombinedDiff(string line)
-    {
-        return line.StartsWith("--") || line.StartsWith("- ") || line.StartsWith(" -");
+                    // start may be intended, if a new block of changes starts with white spaces
+                    && (textMarkers[0].Offset <= textOffset || (firstNonWhiteSpace > 0 && textMarkers[0].Offset <= textOffset + firstNonWhiteSpace))
+
+                    // Compensate length->ending and remove the trailing newline chars (no check for \r\n vs \n)
+                    && (textMarkers[0].EndOffset >= textOffset + textLength - 3)
+
+                    // Assume the user has not overridden colors
+                    && MarkerColorMatch(textMarkers[0], lineType);
+        }
+
+        static bool MarkerColorMatch(TextMarker textMarker, DiffLineType lineType)
+        {
+            // The expected marker color for a line type, for heuristics.
+
+            return lineType is (DiffLineType.Minus or DiffLineType.MinusLeft)
+                ? (AppSettings.ReverseGitColoring.Value
+                    ? textMarker.Color == AppColor.AnsiTerminalRedBackNormal.GetThemeColor()
+                    : textMarker.ForeColor == AppColor.AnsiTerminalRedForeNormal.GetThemeColor())
+                : (AppSettings.ReverseGitColoring.Value
+                    ? textMarker.Color == AppColor.AnsiTerminalGreenBackNormal.GetThemeColor()
+                    : textMarker.ForeColor == AppColor.AnsiTerminalGreenForeNormal.GetThemeColor());
```

Comment:
Consider reading `AppSettings.ReverseGitColoring.Value` at most once per document.

Reply:
What is the cost? This is used in other parts too.

Reply:
Similar could be done in the Difftastic highlighter, other seem OK

Reply:
~1000-4000 settings evaluations take 1 ms. This can add up.

Reply:
Cache at begining of process. Store colors in something like a dictionary of some key and the color value.  Just make sure dictionary is reloaded if settings change or even if new analyzer created.